File: ios-native/Jottery/Jottery/Crypto/CryptoService.swift
import CryptoKit
import CommonCrypto
import Foundation

/// Encrypted data wire format — JSON: {"ciphertext":"<base64>","iv":"<base64>"}
/// Base64 of ciphertext includes the GCM auth tag (last 16 bytes).
struct EncryptedData: Codable, Equatable {
    let ciphertext: String  // base64(ciphertext + tag)
    let iv: String          // base64(12-byte nonce)
}

/// AES-256-GCM encryption service, byte-for-byte compatible with the web and TUI clients.
///
/// Wire format matches `src/lib/services/crypto.ts` and `tui/src/crypto/service.rs`:
/// - Key derivation: PBKDF2-HMAC-SHA256
/// - Cipher: AES-256-GCM (CryptoKit)
/// - Nonce: 12 random bytes
/// - Ciphertext encoding: base64(ciphertext || 16-byte tag)
/// - JSON: `{"ciphertext":"...","iv":"..."}`
enum CryptoService {

    // MARK: - Constants

    static let keyLength = 32         // 256 bits
    static let nonceLength = 12       // 96 bits — GCM recommendation
    static let saltLength = 32        // 256 bits
    static let defaultIterations: UInt32 = 600_000
    static let minIterations: UInt32 = 100_000
    static let maxHashChainLength = 50

    // MARK: - Key Derivation

    /// Derive a 256-bit AES key from a password using PBKDF2-HMAC-SHA256.
    /// Uses CommonCrypto (`CCKeyDerivationPBKDF`) for compatibility.
    static func deriveKey(password: String, salt: Data, iterations: UInt32) -> SymmetricKey {
        let effectiveIterations = max(iterations, minIterations)
        let passwordData = Array(password.utf8)

        var derivedKey = Data(count: keyLength)
        _ = derivedKey.withUnsafeMutableBytes { derivedKeyPtr in
            salt.withUnsafeBytes { saltPtr in
                CCKeyDerivationPBKDF(
                    CCPBKDFAlgorithm(kCCPBKDF2),
                    passwordData,
                    passwordData.count,
                    saltPtr.bindMemory(to: UInt8.self).baseAddress!,
                    salt.count,
                    CCPseudoRandomAlgorithm(kCCPRFHmacAlgSHA256),
                    effectiveIterations,
                    derivedKeyPtr.bindMemory(to: UInt8.self).baseAddress!,
                    keyLength
                )
            }
        }
        return SymmetricKey(data: derivedKey)
    }

    // MARK: - Encrypt / Decrypt

    /// Encrypt plaintext to the standard wire format.
    static func encryptText(_ plaintext: String, key: SymmetricKey) throws -> EncryptedData {
        let data = Data(plaintext.utf8)
        return try encrypt(data, key: key)
    }

    /// Decrypt the standard wire format back to plaintext.
    static func decryptText(_ encrypted: EncryptedData, key: SymmetricKey) throws -> String {
        let data = try decrypt(encrypted, key: key)
        guard let text = String(data: data, encoding: .utf8) else {
            throw CryptoError.invalidUTF8
        }
        return text
    }

    /// Encrypt raw data.
    static func encrypt(_ data: Data, key: SymmetricKey) throws -> EncryptedData {
        let nonce = AES.GCM.Nonce()
        let sealed = try AES.GCM.seal(data, using: key, nonce: nonce)

        // CryptoKit puts ciphertext and tag in separate properties.
        // Wire format expects ciphertext || tag concatenated then base64-encoded.
        var combined = sealed.ciphertext
        combined.append(sealed.tag)

        return EncryptedData(
            ciphertext: combined.base64EncodedString(),
            iv: Data(nonce).base64EncodedString()
        )
    }

    /// Decrypt raw data.
    static func decrypt(_ encrypted: EncryptedData, key: SymmetricKey) throws -> Data {
        guard let combinedData = Data(base64Encoded: encrypted.ciphertext),
              let nonceData = Data(base64Encoded: encrypted.iv) else {
            throw CryptoError.invalidBase64
        }

        guard combinedData.count >= 16 else {
            throw CryptoError.ciphertextTooShort
        }

        let ciphertext = combinedData.prefix(combinedData.count - 16)
        let tag = combinedData.suffix(16)

        let nonce = try AES.GCM.Nonce(data: nonceData)
        let sealedBox = try AES.GCM.SealedBox(nonce: nonce, ciphertext: ciphertext, tag: tag)
        return try AES.GCM.open(sealedBox, using: key)
    }

    // MARK: - JSON Helpers

    /// Encrypt a Codable value as JSON.
    static func encryptJSON<T: Encodable>(_ value: T, key: SymmetricKey) throws -> EncryptedData {
        let data = try JSONEncoder().encode(value)
        guard let json = String(data: data, encoding: .utf8) else {
            throw CryptoError.encodingFailed
        }
        return try encryptText(json, key: key)
    }

    /// Decrypt JSON back to a Codable value.
    static func decryptJSON<T: Decodable>(_ encrypted: EncryptedData, key: SymmetricKey) throws -> T {
        let json = try decryptText(encrypted, key: key)
        guard let data = json.data(using: .utf8) else {
            throw CryptoError.invalidUTF8
        }
        return try JSONDecoder().decode(T.self, from: data)
    }

    /// Encrypt a string array (e.g. tags) — JSON-stringify then encrypt.
    static func encryptStringArray(_ strings: [String], key: SymmetricKey) throws -> EncryptedData {
        return try encryptJSON(strings, key: key)
    }

    /// Decrypt a string array.
    static func decryptStringArray(_ encrypted: EncryptedData, key: SymmetricKey) throws -> [String] {
        return try decryptJSON(encrypted, key: key)
    }

    // MARK: - Hashing

    /// SHA-256 hash, returned as standard base64.
    static func hash(_ data: String) -> String {
        let digest = SHA256.hash(data: Data(data.utf8))
        return Data(digest).base64EncodedString()
    }

    /// Compute content hash for conflict detection.
    /// Must match the web client's `computeContentHash` exactly.
    static func computeContentHash(for note: SyncableNoteFields) -> String {
        // Build deterministic JSON matching web client structure.
        // Use ordered keys via manual string building to match JSON.stringify output.
        var attachmentIds = note.attachmentIds
        attachmentIds.sort()

        let jsonParts: [(String, Any)] = [
            ("archived", note.archived),
            ("attachments", attachmentIds),
            ("content", note.content),
            ("locked", note.locked),
            ("pinned", note.pinned),
            ("showPreview", note.showPreview),
            ("syntaxLanguage", note.syntaxLanguage as Any),
            ("tags", note.tags),
            ("wordWrap", note.wordWrap),
        ]

        // JSON.stringify produces alphabetically ordered keys for simple objects.
        // We need to match that exactly.
        let json = buildDeterministicJSON(jsonParts)
        return hash(json)
    }

    // MARK: - Hash Chain

    /// Update hash chain by prepending new hash and trimming.
    static func updateHashChain(currentHash: String, parentChain: [String] = []) -> [String] {
        var newChain = [currentHash] + parentChain
        if newChain.count > maxHashChainLength {
            newChain = Array(newChain.prefix(maxHashChainLength))
        }
        return newChain
    }

    /// Find common ancestor between two hash chains.
    static func findCommonAncestor(chainA: [String], chainB: [String]) -> String? {
        let chainBSet = Set(chainB)
        for hash in chainA {
            if chainBSet.contains(hash) {
                return hash
            }
        }
        return nil
    }

    // MARK: - Random Generation

    /// Generate random salt (32 bytes).
    static func generateSalt() -> Data {
        var salt = Data(count: saltLength)
        salt.withUnsafeMutableBytes { ptr in
            _ = SecRandomCopyBytes(kSecRandomDefault, saltLength, ptr.baseAddress!)
        }
        return salt
    }

    /// Generate UUID v4 string.
    static func generateUUID() -> String {
        UUID().uuidString.lowercased()
    }

    // MARK: - Wire Format Helpers

    /// Parse encrypted data from the JSON string stored in the database.
    /// Format: `{"ciphertext":"...","iv":"..."}`
    static func parseEncryptedJSON(_ jsonString: String) throws -> EncryptedData {
        guard let data = jsonString.data(using: .utf8) else {
            throw CryptoError.invalidUTF8
        }
        return try JSONDecoder().decode(EncryptedData.self, from: data)
    }

    /// Serialize encrypted data to JSON string for storage.
    static func serializeEncryptedJSON(_ encrypted: EncryptedData) throws -> String {
        let data = try JSONEncoder().encode(encrypted)
        guard let json = String(data: data, encoding: .utf8) else {
            throw CryptoError.encodingFailed
        }
        return json
    }
}

// MARK: - Supporting Types

/// Fields needed for content hash computation.
struct SyncableNoteFields {
    let content: String         // Encrypted content string
    let tags: [String]          // Encrypted tags array
    let attachmentIds: [String]
    let pinned: Bool
    let archived: Bool
    let locked: Bool
    let syntaxLanguage: String?
    let wordWrap: Bool
    let showPreview: Bool
    let color: String?
}

enum CryptoError: LocalizedError {
    case invalidBase64
    case ciphertextTooShort
    case invalidUTF8
    case encodingFailed
    case decryptionFailed

    var errorDescription: String? {
        switch self {
        case .invalidBase64: return "Invalid base64 data"
        case .ciphertextTooShort: return "Ciphertext too short"
        case .invalidUTF8: return "Invalid UTF-8 data"
        case .encodingFailed: return "Encoding failed"
        case .decryptionFailed: return "Decryption failed. Invalid key or corrupted data."
        }
    }
}

// MARK: - Deterministic JSON Builder

/// Build a JSON string with keys in the exact order provided.
/// Matches `JSON.stringify` output for the content hash computation.
private func buildDeterministicJSON(_ pairs: [(String, Any)]) -> String {
    var parts: [String] = []
    for (key, value) in pairs {
        let encodedKey = jsonEncode(key)
        let encodedValue = jsonEncodeValue(value)
        parts.append("\(encodedKey):\(encodedValue)")
    }
    return "{\(parts.joined(separator: ","))}"
}

private func jsonEncode(_ string: String) -> String {
    let escaped = string
        .replacingOccurrences(of: "\\", with: "\\\\")
        .replacingOccurrences(of: "\"", with: "\\\"")
        .replacingOccurrences(of: "\n", with: "\\n")
        .replacingOccurrences(of: "\r", with: "\\r")
        .replacingOccurrences(of: "\t", with: "\\t")
    return "\"\(escaped)\""
}

private func jsonEncodeValue(_ value: Any) -> String {
    switch value {
    case let b as Bool:
        return b ? "true" : "false"
    case let s as String:
        return jsonEncode(s)
    case let arr as [String]:
        let items = arr.map { jsonEncode($0) }.joined(separator: ",")
        return "[\(items)]"
    case is NSNull:
        return "null"
    case let opt as String?:
        if let s = opt {
            return jsonEncode(s)
        }
        return "null"
    default:
        // For optional types that are nil
        if case Optional<Any>.none = value {
            return "null"
        }
        return jsonEncode(String(describing: value))
    }
}

File: android-native/app/src/main/java/com/jottery/android/crypto/CryptoService.kt
package com.jottery.android.crypto

import android.util.Base64
import com.jottery.android.model.SyncableNoteFields
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import java.security.SecureRandom
import java.util.UUID
import javax.crypto.Cipher
import javax.crypto.SecretKey
import javax.crypto.SecretKeyFactory
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.PBEKeySpec
import javax.crypto.spec.SecretKeySpec

/**
 * AES-256-GCM encryption service, byte-for-byte compatible with the iOS, web, and TUI clients.
 *
 * Wire format matches iOS `CryptoService.swift`:
 * - Key derivation: PBKDF2-HMAC-SHA256
 * - Cipher: AES-256-GCM
 * - Nonce: 12 random bytes
 * - Ciphertext encoding: base64(ciphertext || 16-byte tag)
 * - JSON: {"ciphertext":"...","iv":"..."}
 *
 * Note: Java's AES/GCM/NoPadding automatically appends the 16-byte GCM tag
 * to the ciphertext during encryption, and expects it appended during decryption.
 * This matches iOS's manual concatenation of ciphertext + tag.
 */
object CryptoService {

    // MARK: - Constants

    const val KEY_LENGTH = 32           // 256 bits
    const val NONCE_LENGTH = 12         // 96 bits — GCM recommendation
    const val SALT_LENGTH = 32          // 256 bits
    const val TAG_LENGTH_BITS = 128     // 16 bytes
    const val DEFAULT_ITERATIONS = 600_000
    const val MIN_ITERATIONS = 100_000
    const val MAX_HASH_CHAIN_LENGTH = 50

    private val secureRandom = SecureRandom()
    private val json = Json { ignoreUnknownKeys = true }

    // MARK: - Key Derivation

    /**
     * Derive a 256-bit AES key from a password using PBKDF2-HMAC-SHA256.
     * Uses the same algorithm as iOS's CCKeyDerivationPBKDF.
     *
     * IMPORTANT: PBEKeySpec takes a CharArray. The PBKDF2 standard uses the
     * UTF-8 encoding of the password, and Java's PBKDF2WithHmacSHA256 does this
     * correctly by converting the CharArray to UTF-8 bytes internally.
     */
    fun deriveKey(password: String, salt: ByteArray, iterations: Int): SecretKey {
        val effectiveIterations = maxOf(iterations, MIN_ITERATIONS)
        val spec = PBEKeySpec(
            password.toCharArray(),
            salt,
            effectiveIterations,
            KEY_LENGTH * 8,  // key length in bits
        )
        val factory = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA256")
        val keyBytes = factory.generateSecret(spec).encoded
        spec.clearPassword()
        return SecretKeySpec(keyBytes, "AES")
    }

    // MARK: - Encrypt / Decrypt

    /** Encrypt plaintext to the standard wire format. */
    fun encryptText(plaintext: String, key: SecretKey): EncryptedData {
        return encrypt(plaintext.toByteArray(Charsets.UTF_8), key)
    }

    /** Decrypt the standard wire format back to plaintext. */
    fun decryptText(encrypted: EncryptedData, key: SecretKey): String {
        val data = decrypt(encrypted, key)
        return String(data, Charsets.UTF_8)
    }

    /** Encrypt raw data. */
    fun encrypt(data: ByteArray, key: SecretKey): EncryptedData {
        val nonce = ByteArray(NONCE_LENGTH)
        secureRandom.nextBytes(nonce)

        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        val gcmSpec = GCMParameterSpec(TAG_LENGTH_BITS, nonce)
        cipher.init(Cipher.ENCRYPT_MODE, key, gcmSpec)

        // Java GCM auto-appends the 16-byte tag to the ciphertext output.
        // This matches iOS's manual ciphertext + tag concatenation.
        val combined = cipher.doFinal(data)

        return EncryptedData(
            ciphertext = Base64.encodeToString(combined, Base64.NO_WRAP),
            iv = Base64.encodeToString(nonce, Base64.NO_WRAP),
        )
    }

    /** Decrypt raw data. */
    fun decrypt(encrypted: EncryptedData, key: SecretKey): ByteArray {
        val combinedData = Base64.decode(encrypted.ciphertext, Base64.NO_WRAP)
        val nonceData = Base64.decode(encrypted.iv, Base64.NO_WRAP)

        require(combinedData.size >= 16) { "Ciphertext too short" }

        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        val gcmSpec = GCMParameterSpec(TAG_LENGTH_BITS, nonceData)
        cipher.init(Cipher.DECRYPT_MODE, key, gcmSpec)

        // Java expects ciphertext + tag appended (same as our wire format).
        return cipher.doFinal(combinedData)
    }

    // MARK: - JSON Helpers

    /** Encrypt a JSON-serialisable value. */
    inline fun <reified T> encryptJSON(value: T, key: SecretKey): EncryptedData {
        val jsonStr = json.encodeToString(value)
        return encryptText(jsonStr, key)
    }

    /** Decrypt JSON back to a typed value. */
    inline fun <reified T> decryptJSON(encrypted: EncryptedData, key: SecretKey): T {
        val jsonStr = decryptText(encrypted, key)
        return json.decodeFromString(jsonStr)
    }

    /** Encrypt a string array (e.g. tags) — JSON-stringify then encrypt. */
    fun encryptStringArray(strings: List<String>, key: SecretKey): EncryptedData {
        return encryptJSON(strings, key)
    }

    /** Decrypt a string array. */
    fun decryptStringArray(encrypted: EncryptedData, key: SecretKey): List<String> {
        return decryptJSON(encrypted, key)
    }

    // MARK: - Hashing

    /** SHA-256 hash, returned as standard base64. */
    fun hash(data: String): String {
        val digest = java.security.MessageDigest.getInstance("SHA-256")
        val hashBytes = digest.digest(data.toByteArray(Charsets.UTF_8))
        return Base64.encodeToString(hashBytes, Base64.NO_WRAP)
    }

    /**
     * Compute content hash for conflict detection.
     * Must match the web client's computeContentHash exactly.
     */
    fun computeContentHash(note: SyncableNoteFields): String {
        val sortedAttachmentIds = note.attachmentIds.sorted()

        val pairs = listOf(
            "archived" to jsonEncodeValue(note.archived),
            "attachments" to jsonEncodeArray(sortedAttachmentIds),
            "content" to jsonEncode(note.content),
            "locked" to jsonEncodeValue(note.locked),
            "pinned" to jsonEncodeValue(note.pinned),
            "showPreview" to jsonEncodeValue(note.showPreview),
            "syntaxLanguage" to jsonEncodeNullableString(note.syntaxLanguage),
            "tags" to jsonEncodeArray(note.tags),
            "wordWrap" to jsonEncodeValue(note.wordWrap),
        )

        val json = "{" + pairs.joinToString(",") { (k, v) -> "${jsonEncode(k)}:$v" } + "}"
        return hash(json)
    }

    // MARK: - Hash Chain

    /** Update hash chain by prepending new hash and trimming. */
    fun updateHashChain(currentHash: String, parentChain: List<String> = emptyList()): List<String> {
        val newChain = mutableListOf(currentHash)
        newChain.addAll(parentChain)
        return if (newChain.size > MAX_HASH_CHAIN_LENGTH) {
            newChain.take(MAX_HASH_CHAIN_LENGTH)
        } else {
            newChain
        }
    }

    /** Find common ancestor between two hash chains. */
    fun findCommonAncestor(chainA: List<String>, chainB: List<String>): String? {
        val chainBSet = chainB.toSet()
        for (hash in chainA) {
            if (hash in chainBSet) return hash
        }
        return null
    }

    // MARK: - Random Generation

    /** Generate random salt (32 bytes). */
    fun generateSalt(): ByteArray {
        val salt = ByteArray(SALT_LENGTH)
        secureRandom.nextBytes(salt)
        return salt
    }

    /** Generate UUID v4 string. */
    fun generateUUID(): String = UUID.randomUUID().toString().lowercase()

    // MARK: - Wire Format Helpers

    /** Parse encrypted data from the JSON string stored in the database. */
    fun parseEncryptedJSON(jsonString: String): EncryptedData {
        return json.decodeFromString(jsonString)
    }

    /** Serialise encrypted data to JSON string for storage. */
    fun serializeEncryptedJSON(encrypted: EncryptedData): String {
        return json.encodeToString(encrypted)
    }

    // MARK: - Deterministic JSON Builder (private)

    private fun jsonEncode(string: String): String {
        val escaped = string
            .replace("\\", "\\\\")
            .replace("\"", "\\\"")
            .replace("\n", "\\n")
            .replace("\r", "\\r")
            .replace("\t", "\\t")
        return "\"$escaped\""
    }

    private fun jsonEncodeValue(value: Boolean): String = if (value) "true" else "false"

    private fun jsonEncodeNullableString(value: String?): String {
        return if (value != null) jsonEncode(value) else "null"
    }

    private fun jsonEncodeArray(items: List<String>): String {
        return "[" + items.joinToString(",") { jsonEncode(it) } + "]"
    }
}

